packages/assistify-ai/server/methods/SmartiWidgetBackend.js

```diff
@@ -185,3 +186,51 @@ Meteor.methods({
 		return script;
 	}
 });
+
+Meteor.methods({
+	triggerFullResync() {
+		const requests = RocketChat.models.Rooms.model.find({t:'r', closer: {$exists: false}}).fetch();
+		SystemLogger.debug('Number of Requests to sync: ', requests.length);
+		for (let i=0; i < requests.length; i++) {
+			Meteor.call('tryResync', requests[i]._id);
+		}
+		const topics = RocketChat.models.Rooms.model.find({t:'e', closer: {$exists: false}}).fetch();
+		SystemLogger.debug('Number of Topics to sync: ', topics.length);
+		for (let i=0; i < topics.length; i++) {
+			Meteor.call('tryResync', topics[i]._id);
+		}
+	}
+});
+
+Meteor.methods({
+	markMessageAsSynced(messageId) {
+		const messageDB = RocketChat.models.Messages;
+		const message = messageDB.findOneById(messageId);
+		const lastUpdate = message ? message._updatedAt : 0;
+		if (lastUpdate) {
+			messageDB.model.update(
+				{_id: messageId},
+				{
+					$set: {
+						lastSync: lastUpdate
+					}
+				});
+			SystemLogger.debug('Message Id: ', messageId, ' has been synced');
+		} else {
+			SystemLogger.debug('Message Id: ', messageId, ' can not be synced');
+		}
+	}
+});
+
+Meteor.methods({
+	tryResync(rid) {
+		SystemLogger.debug('Sync all unsynced messages in room: ', rid);
+		const knowledgeAdapter = getKnowledgeAdapter();
+		const messageDB = RocketChat.models.Messages;
+		const messages = messageDB.find({ lastSync: { $exists: false }, rid }).fetch();
+		for (let i=0; i < messages.length; i++) {
+			console.log(messages[i]._id);
+			knowledgeAdapter.onMessage(messages[i]);
```

Comment:
Passing each message individually has a veeeeeeery poor performance. 
This mostly is due to the fact that an analysis is getting triggered after each message. 
I'd expect to pass always the full conversation. The adapter might need a method for that. 
In the implementation of the method, the messages should be selected and can also be brought into the proper order.